### PR TITLE
UCP: Add pool of registered memory to worker

### PR DESCRIPTION
**Pull request by brminich**

### CODE REVIEW

**Comment by shamisp** [comment]
why ?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1879/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3845/ for details (Mellanox internal link).

File: src/ucp/core/ucp_mm.c [hunk 1]
```diff
@@ -444,3 +445,52 @@ ucp_mem_advise(ucp_context_h context, ucp_mem_h memh,
...
+    ucs_status_t status;
+
+    mem_map_params.field_mask = UCP_MEM_MAP_PARAM_FIELD_ADDRESS |
+                                UCP_MEM_MAP_PARAM_FIELD_LENGTH |
+                                UCP_MEM_MAP_PARAM_FIELD_FLAGS;
+    mem_map_params.address    = NULL;
+    mem_map_params.length     = *size_p + sizeof(*chunk_hdr);
+    mem_map_params.flags      = UCP_MEM_MAP_ALLOCATE;
+
+    status = ucp_mem_map(worker->context, &mem_map_params, &memh);
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c]: how about using internal function (ucp_mem_alloc) instead of API function?

File: src/ucp/core/ucp_worker.c [hunk 2]
```diff
@@ -551,29 +561,36 @@ static void ucp_worker_init_atomic_tls(ucp_worker_h worker)
...
+        max_mp_entry_size = ucs_max(max_mp_entry_size,
+                                    if_attr->cap.am.max_zcopy);
     }
 
-    max_am_mp_entry_size += rx_headroom;
+    status = ucs_mpool_init(&worker->am_mp, 0,
+                            max_mp_entry_size + rx_headroom,
+                            0, UCS_SYS_CACHE_LINE_SIZE, 128, UINT_MAX,
+                            &ucp_am_mpool_ops, "ucp_am_bufs");
+    if (status == UCS_OK) {
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: pls use "goto/return" for err handling like the rest of the code

**Comment by yosefe** [comment]
@shamisp It's needed for optimizing zero-copy operations when memory registration/cache lookup add overhead. For example, IOV operations and posting tags to offload. In these cases UCT requires zero-copy.
Also, it would be used to implement pipeline copy algorithm for GPU memory.

File: src/ucp/core/ucp_mm.c (hunk 1 above)
Comment (brminich) [inline src/ucp/core/ucp_mm.c]: In this case I'd need to allocate memh by my own, but seems I'd need to do that, because mem_map takes ctx lock.

Comment (yosefe) [inline src/ucp/core/ucp_mm.c]: maybe it's better this way, and also allocate "internal" structure (non ucp_mem_h) instead?

**Comment by shamisp** [comment]
@yosefe - this justification has to included in the commit

**Comment by brminich** [comment]
@yosefe, comments addressed in the second commit. Still use ucp_mem_h, considered other approaches, but they look worse (more hacky) due to using of flexible array.
Only commit message is changed in the first commit (according to @shamisp comment)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1898/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1899/ for details.

**Comment by yosefe** [comment]
@brminich why need to define ucp_men_[un]map_common and not just call ucp_mem_alloc?

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3871/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
What is the idea - register the memory with all memory domains ?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3882/ for details (Mellanox internal link).

**Comment by brminich** [comment]
@shamisp, yes this was an intention

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3888/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3890/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3897/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3908/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp pls review as well

**Comment by shamisp** [comment]
👍

**Pull request merged**